snuba/snuba_migrations/spans/0001_spans_v1.py

```diff
@@ -0,0 +1,99 @@
+from typing import List, Sequence
+
+from snuba.clickhouse.columns import UUID, Column, DateTime, Nested, String, UInt
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.operations import OperationTarget, SqlOperation
+from snuba.utils.schemas import Float
+
+storage_set_name = StorageSetKey.SPANS
+local_table_name = "spans_local"
+dist_table_name = "spans_dist"
+
+UNKNOWN_SPAN_STATUS = 2
+
+columns: List[Column[Modifiers]] = [
+    Column("project_id", UInt(64)),
+    Column("transaction_id", UUID(Modifiers(nullable=True))),
+    Column("transaction_op", String(Modifiers(nullable=True))),
+    Column("trace_id", UUID()),
+    Column("span_id", UInt(64)),
+    Column("parent_span_id", UInt(64, Modifiers(nullable=True))),
+    Column("segment_id", UInt(64)),
+    Column("is_segment", UInt(8)),
+    Column("segment_name", String(Modifiers(default="''"))),
+    Column("start_timestamp", DateTime()),
+    Column("end_timestamp", DateTime(Modifiers(codecs=["DoubleDelta"]))),
+    Column("duration", UInt(32)),
+    Column("exclusive_time", Float(64)),
+    Column("op", String(Modifiers(low_cardinality=True))),
+    Column("group", UInt(64)),
+    Column("span_status", UInt(8)),
+    Column("span_kind", String(Modifiers(low_cardinality=True))),
+    Column("description", String()),
+    Column("status", UInt(32, Modifiers(nullable=True))),
+    Column("module", String(Modifiers(low_cardinality=True))),
+    Column("action", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("domain", String(Modifiers(nullable=True))),
+    Column("platform", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("user", String(Modifiers(nullable=True))),
+    Column("tags", Nested([("key", String()), ("value", String())])),
+    Column(
+        "measurements",
+        Nested(
+            [("key", String(Modifiers(low_cardinality=True))), ("value", Float(64))]
+        ),
+    ),
+    Column("partition", UInt(16)),
+    Column("offset", UInt(64)),
+    Column("retention_days", UInt(16)),
+    Column("deleted", UInt(8)),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=storage_set_name,
+                table_name=local_table_name,
+                columns=columns,
+                engine=table_engines.ReplacingMergeTree(
+                    order_by="(project_id, group, end_timestamp, cityHash64(span_id))",
```

Comment:
Does the `group` have a 1:1 relationship with a segment_name or with segment_id ? 
I think this was discussed in the draft, but I cannot find that PR anymore. Any change we can sort spans by transaction (or segment_id) within a time range? 

```
..... toStartOfHour(end_timestamp), segment_id, cityHash64(span_id)
```

This is meant to pack together the spans for a segment_id thus achieving a much better compression as data will be more homogeneous on very high entropy columns like transaction id and segment id

Reply:
The `group` does not bear any relationship with a segment_name or with a `segment_id`. `group` is like a hash of some common elements of a span (maybe like module, op, etc.). Different transactions (with different transaction_id but same transaction_name) would have the same common elements of the span. But it is also possible that the same common elements of the span can belong to completely different transactions. For example, a db call to get user count from a table can be triggered from different workflows and hence different transaction names. 

With `group` being the primary entity by which individual samples are found, the spans belonging to an individual segment_id are bound to be spread across in disk. My first proposal was aligned to have all spans belonging to a segment be close to each other. But the product workflow (see https://github.com/getsentry/snuba/pull/4089#issuecomment-1527056783) points towards use cases which led me to make group be so high in the sorting order.

Reply:
Ok, we will have to watch out very carefully for disk usage then as, spreading all those segment/transactions unique ids will hit us on compression efficiency quite hard.